File: mal/impls/es6/reader.mjs
import { _keyword, _assoc_BANG, Vector } from './types';

export class BlankException extends Error {}

class Reader {
    constructor(tokens) {
        this.tokens = tokens
        this.position = 0
    }
    next() { return this.tokens[this.position++] }
    peek() { return this.tokens[this.position] }
}

function tokenize(str) {
    const re = /[\s,]*(~@|[\[\]{}()'`~^@]|"(?:\\.|[^\\"])*"?|;.*|[^\s\[\]{}('"`,;)]*)/g
    let match = null
    let results = []
    while ((match = re.exec(str)[1]) != '') {
        if (match[0] === ';') { continue }
        results.push(match)
    }
    return results
}

function read_atom (reader) {
    const token = reader.next()
    //console.log("read_atom:", token)
    if (token.match(/^-?[0-9]+$/)) {
        return parseInt(token,10)        // integer
    } else if (token.match(/^-?[0-9][0-9.]*$/)) {
        return parseFloat(token,10)     // float
    } else if (token.match(/^"(?:\\.|[^\\"])*"$/)) {
        return token.slice(1,token.length-1)
            .replace(/\\(.)/g, (_, c) => c === "n" ? "\n" : c)
    } else if (token[0] === "\"") {
        throw new Error("expected '\"', got EOF");
    } else if (token[0] === ":") {
        return _keyword(token.slice(1))
    } else if (token === "nil") {
        return null
    } else if (token === "true") {
        return true
    } else if (token === "false") {
        return false
    } else {
        return Symbol.for(token) // symbol
    }
}

// read list of tokens
function read_list(reader, start, end) {
    start = start || '('
    end = end || ')'
    var ast = []
    var token = reader.next()
    if (token !== start) {
        throw new Error("expected '" + start + "'")
    }
    while ((token = reader.peek()) !== end) {
        if (!token) {
            throw new Error("expected '" + end + "', got EOF")
        }
        ast.push(read_form(reader))
    }
    reader.next()
    return ast
}

// read vector of tokens
function read_vector(reader) {
    return Vector.from(read_list(reader, '[', ']'));
}

// read hash-map key/value pairs
function read_hash_map(reader) {
    return _assoc_BANG(new Map(), ...read_list(reader, '{', '}'))
}

function read_form(reader) {
    var token = reader.peek()
    switch (token) {
    // reader macros/transforms
    case ';': return null // Ignore comments
    case '\'': reader.next()
               return [Symbol.for('quote'), read_form(reader)]
    case '`': reader.next()
              return [Symbol.for('quasiquote'), read_form(reader)]
    case '~': reader.next()
              return [Symbol.for('unquote'), read_form(reader)]
    case '~@': reader.next()
               return [Symbol.for('splice-unquote'), read_form(reader)]
    case '^': reader.next()
              var meta = read_form(reader)
              return [Symbol.for('with-meta'), read_form(reader), meta]
    case '@': reader.next()
              return [Symbol.for('deref'), read_form(reader)]

    // list
    case ')': throw new Error("unexpected ')'")
    case '(': return read_list(reader)

    // vector
    case ']': throw new Error("unexpected ']'")
    case '[': return read_vector(reader)

    // hash-map
    case '}': throw new Error("unexpected '}'")
    case '{': return read_hash_map(reader)

    // atom
    default:  return read_atom(reader)
    }
}

export function read_str(str) {
    var tokens = tokenize(str)
    if (tokens.length === 0) { throw new BlankException() }
    return read_form(new Reader(tokens))
}

File: mal/impls/nim/reader.nim
import options, re, strutils, types

let
  tokenRE = re"""[\s,]*(~@|[\[\]{}()'`~^@]|"(?:\\.|[^\\"])*"?|;.*|[^\s\[\]{}('"`,;)]*)"""
  intRE   = re"-?[0-9]+$"
  strRE   = re"""^"(?:\\.|[^\\"])*"$"""

type
  Blank* = object of Exception

  Reader = object
    tokens: seq[string]
    position: int

proc next(r: var Reader): Option[string] =
  if r.position < r.tokens.len:
    result = r.tokens[r.position].some
    inc r.position

proc peek(r: Reader): Option[string] =
  if r.position < r.tokens.len: return r.tokens[r.position].some

proc tokenize(str: string): seq[string] =
  result = @[]
  var pos = 0
  while pos < str.len:
    var matches: array[2, string]
    var len = str.findBounds(tokenRE, matches, pos)
    if len.first != -1 and len.last != -1 and len.last >= len.first:
      pos = len.last + 1
      if matches[0].len > 0 and matches[0][0] != ';':
        result.add matches[0]
    else:
      inc pos

proc read_form(r: var Reader): MalType

proc read_seq(r: var Reader, fr, to: string): seq[MalType] =
  result = @[]
  var t = r.next
  if t.get("") != fr: raise newException(ValueError, "expected '" & fr & "'")

  t = r.peek
  while t.get("") != to:
    if t.get("") == "": raise newException(ValueError, "expected '" & to & "', got EOF")
    result.add r.read_form
    t = r.peek
  discard r.next

proc read_list(r: var Reader): MalType =
  result = list r.read_seq("(", ")")

proc read_vector(r: var Reader): MalType =
  result = vector r.read_seq("[", "]")

proc read_hash_map(r: var Reader): MalType =
  result = hash_map r.read_seq("{", "}")

proc read_atom(r: var Reader): MalType =
  let t = r.next.get("")
  if t.match(intRE): number t.parseInt
  elif t[0] == '"':
    if not t.match(strRE):
      raise newException(ValueError, "expected '\"', got EOF")
    str t[1 ..< t.high].multiReplace(("\\\"", "\""), ("\\n", "\n"), ("\\\\", "\\"))
  elif t[0] == ':':  keyword t[1 .. t.high]
  elif t == "nil":   nilObj
  elif t == "true":  trueObj
  elif t == "false": falseObj
  else:              symbol t

proc read_form(r: var Reader): MalType =
  if r.peek.get("")[0] == ';':
    discard r.next
    return nilObj
  case r.peek.get("")
  of "'":
    discard r.next
    result = list(symbol "quote", r.read_form)
  of "`":
    discard r.next
    result = list(symbol "quasiquote", r.read_form)
  of "~":
    discard r.next
    result = list(symbol "unquote", r.read_form)
  of "~@":
    discard r.next
    result = list(symbol "splice-unquote", r.read_form)
  of "^":
    discard r.next
    let meta = r.read_form
    result = list(symbol "with-meta", r.read_form, meta)
  of "@":
    discard r.next
    result = list(symbol "deref", r.read_form)

  # list
  of "(": result = r.read_list
  of ")": raise newException(ValueError, "unexpected ')'")

  # vector
  of "[": result = r.read_vector
  of "]": raise newException(ValueError, "unexpected ']'")

  # hash-map
  of "{": result = r.read_hash_map
  of "}": raise newException(ValueError, "unexpected '}'")

  # atom
  else:   result = r.read_atom

proc read_str*(str: string): MalType =
  var r = Reader(tokens: str.tokenize)
  if r.tokens.len == 0:
    raise newException(Blank, "Blank line")
  r.read_form
